File: Exercise 8/Java/src/main/java/de/thu/recipebook/activities/RecipeListActivity.java
package de.thu.recipebook.activities;

import android.content.Intent;
import android.os.Bundle;
import android.os.StrictMode;
import android.provider.Settings;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import de.thu.recipebook.R;
import de.thu.recipebook.databases.FavoritesDbHelper;
import de.thu.recipebook.models.Recipe;
import de.thu.recipebook.models.RecipeCollection;
import de.thu.recipebook.models.RecipeListAdapter;
import de.thu.recipebook.runnables.FetchRecipeListRunnable;
import okhttp3.Credentials;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class RecipeListActivity extends AppCompatActivity {
    private RecipeCollection recipeCollection;
    private RecipeListAdapter adapter;
    private FavoritesDbHelper favoritesDbHelper;

    private FetchRecipeListRunnable runnable;

    private ListView listView;
    private EditText editTextSearch;

    private String filterName;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_recipe_list);

        recipeCollection = RecipeCollection.getInstance();
        favoritesDbHelper = new FavoritesDbHelper(this);
        adapter = new RecipeListAdapter(favoritesDbHelper);
        List<Recipe> data = recipeCollection.getRecipes();
        adapter.setData(data);

        runnable = new FetchRecipeListRunnable(this);
        new Thread(runnable).start();

        listView = findViewById(R.id.list_view_recipe);
        listView.setAdapter(adapter);
        listView.setOnItemClickListener((adapterView, view, i, l) -> {
            Intent recipeDetailsIntent = new Intent(RecipeListActivity.this, RecipeDetailsActivity.class);
            recipeDetailsIntent.putExtra("id", recipeCollection.getRecipes().get(i).getId());
            startActivity(recipeDetailsIntent);
        });
        editTextSearch = findViewById(R.id.edit_text_search);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.recipe_list_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.add_recipe_entry:
                Intent createRecipeIntent = new Intent(this, SaveRecipeActivity.class);
                startActivity(createRecipeIntent);
                break;
            case R.id.refresh_entry:
                new Thread(runnable).start();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onResume() {
        adapter.notifyDataSetChanged();
        super.onResume();
    }

    public void searchButtonOnClick(View view) {
        filterName = editTextSearch.getText().toString();
    }

    public void notifyDataSetChanged() {
        adapter.notifyDataSetChanged();
    }

    public String getFilterName() {
        return filterName;
    }
}

File: documentation/Exercise 7/Kotlin/src/main/java/de/thu/recipebook/activities/RecipeListActivity.kt
package de.thu.recipebook.activities

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.EditText
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import de.thu.recipebook.R
import de.thu.recipebook.models.Recipe
import de.thu.recipebook.models.RecipeCollection
import de.thu.recipebook.models.RecipeListAdapter
import de.thu.recipebook.runnables.FetchRecipeListRunnable

class RecipeListActivity : AppCompatActivity() {
    private var recipeCollection: RecipeCollection? = null
    private var adapter: RecipeListAdapter? = null

    private var runnable: FetchRecipeListRunnable? = null

    private var listView: ListView? = null
    private var editTextSearch: EditText? = null

    private var filterName: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_recipe_list)

        recipeCollection = RecipeCollection.instance
        adapter = RecipeListAdapter()
        val data: List<Recipe> = recipeCollection!!.getRecipes()
        adapter!!.setData(data)

        runnable = FetchRecipeListRunnable(this)
        Thread(runnable).start()

        listView = findViewById(R.id.list_view_recipe)
        listView?.adapter = adapter
        listView?.onItemClickListener =
            OnItemClickListener { _: AdapterView<*>?, _: View?, i: Int, _: Long ->
                val recipeDetailsIntent =
                    Intent(this@RecipeListActivity, RecipeDetailsActivity::class.java)
                recipeDetailsIntent.putExtra("id", recipeCollection!!.getRecipes()[i].id)
                startActivity(recipeDetailsIntent)
            }
        editTextSearch = findViewById(R.id.edit_text_search)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.recipe_list_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.add_recipe_entry -> {
                val createRecipeIntent = Intent(this, SaveRecipeActivity::class.java)
                startActivity(createRecipeIntent)
            }
            R.id.refresh_entry -> {
                Thread(runnable).start()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        adapter!!.notifyDataSetChanged()
        super.onResume()
    }

    fun searchButtonOnClick(view: View?) {
        filterName = editTextSearch!!.text.toString()
    }

    fun notifyDataSetChanged() {
        adapter!!.notifyDataSetChanged()
    }

    fun getFilterName(): String? {
        return filterName
    }

}